### ISSUE TITLE
Multiple text column unknown text length

### ISSUE THREAD

**Issue by MichelMichels** [QuestPDF/QuestPDF#387]
**Is your feature request related to a problem? Please describe.**
Is it at the moment possible to have multiple text columns with wrapped text?
See screenshot in additional context.

**Describe alternatives you've considered**
I could measure the length of the text, but this feels janky.

**Additional context**
![image](https://user-images.githubusercontent.com/6505319/198959456-97a2c248-8d83-494d-b4c6-12f4e38efb61.png)

**Comment by MarcinZiabek** [comment]
Hi 😁 

I am afraid that QuestPDF does not offer out-of-the-box support for this type of layout. I am planning to introduce it but it may require some serious architectural changes. 

As an alternative, you can try to utilise the `DynamicComponent` approach but this may be slightly more difficult.

**Comment by MichelMichels** [comment]
No problem. I'll look into DynamicComponent.

Any idea how this could be implemented natively in QuestPDF? Some kind of flow layout? If you need help, I can offer some of my time.

**Comment by MarcinZiabek** [comment]
Yes, I am thinking about this topic for a couple of month already 😁 

Right now, QuestPDF uses the measure-draw approach. That means, the parent can `measure` its children many times, with varying available space provided. And then, once everything is set, the parent can perform the `draw` operation. It draws the child to the canvas and also updates its state (e.g. for the column element, it updates information which items were drawn). 

What is important, the `draw` operation can be (usually) performed only once per page. We need to be able to measure elements with various states. So, in your example, we need to perform the measure-draw operation twice, for the left and right sides of the document. 

If I need to name the process, the current architecture uses 1-step probing, while we need to have many-step probing. 

One of ideas that I have, is to expose elements state. So the parent can save current state of its children, then perform measurement operations, maybe change the state somehow, measure again, etc. This way, we can perform more advanced measurement operations, without actually drawing anything on the canvas. Please notice, that state consists of not only direct children but all descendants. And not all elements have state. This suddenly involves tree traversal, apaplying IDs to elements, using dictionaries to store state, etc. Quite complex stuff.

I am not sure if anything above makes sense 😅 This is a pretty rough idea, and quite difficult to describe. I have done this type of state exposing in the DynamicComponent approach, where the library saves / changes / restores state of the component many times per page drawing step, in order to best predict its size and position.

**Comment by MichelMichels** [comment]
I'm  trying to understand but I think it's hard because I'm not familiar with the internal codebase of QuestPDF 😄. 
I'll look into the DynamicComponent approach and maybe that will help me understand.

**Comment by Azuf** [comment]
@MichelMichels have you tried the `DynamicComponent` approach yet? I have tried the approach of measuring the text and reducing the length till it fits in the container by using the `DyamicContext`.`CreateElement` method but, it's gotten nowhere because I have other requirements that cannot be met using this approach. Maybe you've had better luck with this?

**Comment by MichelMichels** [comment]
Hey @Azuf , I followed another path. I calculated the character count and made 3 columns where I put text in, but there's no automatic wrapping. It adds pages when overflowing the columns. But in my use case, this was not a problem as the content where a lot of paragraphs with titles which didn't need an order to be printed.

So I'm afraid this doesn't help you very much... Anyway, here's the code I used:

```csharp
            container
                .Column(column =>
                {
                    column.Item().Row(row =>
                    {
                        row.RelativeItem()
                            .ShowOnce()
                            .Text(text =>
                            {
                                AddTitledParagraph(text, "title", content);
                                // Statically repeated code above
                            });

                        row
                            .AutoItem()
                            .PaddingHorizontal(5f, Unit.Millimetre)
                            .LineVertical(0.1f)
                            .LineColor(Colors.Grey.Lighten1);

                        row.RelativeItem()
                            .ShowOnce()
                            .Text(text =>
                            {
                                AddTitledParagraph(text, "title", content);
                                // Statically repeated code above
                            });

                        row
                            .AutoItem()
                            .PaddingHorizontal(5f, Unit.Millimetre)
                            .LineVertical(0.1f)
                            .LineColor(Colors.Grey.Lighten1);

                        row.RelativeItem()
                            .ShowOnce()
                            .Text(text =>
                            {
                                AddTitledParagraph(text, "title", content);
                                // Statically repeated code above
                            });
                    });
                });
```

**Comment by MarcinZiabek** [comment]
I'm happy to share that this requirement is now natively supported via the MultiColumn API! 🎉 You can learn more about it here: [MultiColumn API Documentation](https://www.questpdf.com/api-reference/multi-column.html). Hope this helps!

Since there has been no recent activity, I’m closing this issue. If you experience a similar problem or have a related request, feel free to open a new GitHub issue. I appreciate your feedback! ❤️
